src/analysis/functions.rs

```diff
@@ -463,6 +646,112 @@ fn analyze_async(
             success_parameters,
             error_parameters,
         });
+        true
+    } else {
+        false
+    }
+}
+
+fn analyze_callback(
+    func_name: &str,
+    env: &Env,
+    par: &CParameter,
+    callback_info: &Option<CallbackInfo>,
+    commented: &mut bool,
+    imports: &mut Imports,
+    c_parameters: &[(&CParameter, usize)],
+) -> Option<(Trampoline, Option<usize>)> {
+    if let Type::Function(ref func) = env.library.type_(par.typ) {
+        if par.c_type != "GDestroyNotify" {
+            if let Some(user_data) = par.user_data_index {
+                if user_data >= c_parameters.len() {
+                    error!("function `{}` has an invalid user data index of {} when there are {} parameters",
+                           func_name,
+                           user_data,
+                           c_parameters.len());
+                    return None;
+                } else if c_parameters[user_data].0.c_type != "gpointer" {
+                    *commented = true;
+                    error!("function `{}`'s callback `{}` has invalid user data",
+                           func_name,
+                           par.name);
+                    return None;
+                }
+            } else {
+                *commented = true;
+                error!("function `{}`'s callback `{}` without associated user data",
+                       func_name,
+                       par.name);
+                return None;
+            }
+            if let Some(destroy_index) = par.destroy_index {
+                if destroy_index >= c_parameters.len() {
+                    error!("function `{}` has an invalid destroy index of {} when there are {} parameters",
+                           func_name,
+                           destroy_index,
+                           c_parameters.len());
+                    return None;
+                }
+                if c_parameters[destroy_index].0.c_type != "GDestroyNotify" {
+                    *commented = true;
+                    error!("function `{}`'s callback `{}` has invalid destroy callback",
+                           func_name,
+                           par.name);
+                    return None;
+                }
+            }
+        }
+
+        // If we don't have a "user data" parameter, we can't get the closure so there's nothing we
+        // can do...
+        if func.parameters.len() < 1 || func.parameters.last().unwrap().c_type != "gpointer" {
```

Comment:
You already checked for that above, or not? And the `user_data` does not have to be the last parameter, the code above seems to handle other cases too via the indexes just fine

Reply:
The `closure` attribute is not always on the parameter that takes the C function pointer?

Reply:
`func` is the callback **type** here. Not the function.